File: example/sap_listener/sap-listener-example.cc
//
// Copyright (c) 2023, DefenceX PTY LTD
//
// This file is part of the VivoeX project developed by DefenceX.
//
// Licensed under the Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
// License. See the LICENSE file in the project root for full license details.
//
/// \file sap-listener.cc
///

#include <iostream>

#include "sap/sap_listener.h"

int main([[maybe_unused]] int argc, [[maybe_unused]] char* argv[]) {
  std::cout << argv[0] << " starting\n";

  mediax::sap::SAPListener& sap = mediax::sap::SAPListener::GetInstance();

  std::cout << "Waiting for 2 seconds for all SAP/SDP announcements\n";

  // Listen for 2 seconds and dump out the SAP announcements seen
  sap.Start();
  sleep(2);
  sap.Stop();

  const std::map<std::string, mediax::sap::SDPMessage, std::less<>>& announcements = sap.GetSAPAnnouncements();

  if (announcements.empty()) {
    std::cout << "No SAP/SDP announcements seen\n";
    return 0;
  }

  for (const auto& [name, sdp] : announcements) {
    std::cout << "SDP> name: " << sdp.session_name << ", source: " << sdp.ip_address_source
              << ", ipaddr: " << sdp.ip_address << ":" << sdp.port << ", height: " << sdp.height
              << ", width: " << sdp.width << ", framerate: " << sdp.framerate << ", sampling: " << sdp.sampling << "\n";
  }

  return 0;
}

File: example/python/sap_listener_example.py
#!/usr/bin/python3 
#
# Copyright (c) 2023, DefenceX PTY LTD
#
# This file is part of the VivoeX project developed by DefenceX.
#
# Licensed under the Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
# License. See the LICENSE file in the project root for full license details.
#

import time
import sys
sys.path.append('/usr/local/lib/python3/dist-packages/mediax')

import sap_listener
SapListener = sap_listener.SapListener.GetInstance()

SapListener.Start()

print("Waiting for 2 seconds for all SAP/SDP announcements")

# Sleep 2 seconds
time.sleep(2)

SapListener.Stop()

announcements = SapListener.GetSapAnnouncements()

# Dump all discovered announcements.
if not announcements:
    print("No SAP/SDP announcements seen")
    sys.exit(0)

print ("SAP/SDP announcements seen:")
for name, sdp in announcements.items():
    from datetime import datetime
    print(datetime.now().strftime("%Y-%m-%d %H:%M:%S") + ":", end="")

    print(f":{sdp.session_name}:{sdp.ip_address_source}:{sdp.ip_address}:{sdp.port}:{sdp.height}:{sdp.width}:{sdp.framerate}:{sdp.sampling}")
    ## sdp.sdp insert /t after /n to make it more readable
    sdp.sdp_text = sdp.sdp_text.replace("\n", "\n\t")
    print(f"\t{sdp.sdp_text}")